pocs/linux/kernelctf/CVE-2023-52620_lts_cos_mitigation/docs/exploit.md

```diff
@@ -0,0 +1,320 @@
+# Overview
+
+This vulnerability was found in nftables and was patched in June 2023. However, the patch was mistakenly missed in the LTS version and the exploit was available in the LTS kernel for 8 months. Therefore, I reported this to the maintainer so that the patch could be committed to the LTS version.
+
+The vulnerability is caused by being able to set a timeout on an anonymous set. As shown in `nf_tables_newset`, there is no validation of an anonymous set when setting timeout.
+
+```c
+static int nf_tables_newset(struct sk_buff *skb, const struct nfnl_info *info,
+			    const struct nlattr * const nla[])
+{
+	...
+	desc.timeout = 0;
+	if (nla[NFTA_SET_TIMEOUT] != NULL) {
+		if (!(flags & NFT_SET_TIMEOUT))
+			return -EINVAL;
+
+		err = nf_msecs_to_jiffies64(nla[NFTA_SET_TIMEOUT], &desc.timeout);
+		if (err)
+			return err;
+	}
+	...
+```
+
+Binding an anonymous set to `lookup expr` and then deleting expr calls `nf_tables_deactivate_set`. This function calls `nft_map_deactivate` to deactivate objects mapped to elements of the set if it is an anonymous set.
+
+```c
+void nf_tables_deactivate_set(const struct nft_ctx *ctx, struct nft_set *set,
+			      struct nft_set_binding *binding,
+			      enum nft_trans_phase phase)
+{
+	...
+	case NFT_TRANS_PREPARE:
+		if (nft_set_is_anonymous(set)) {
+			if (set->flags & (NFT_SET_MAP | NFT_SET_OBJECT))
+				nft_map_deactivate(ctx, set);
+
+			nft_deactivate_next(ctx->net, set);
+		}
+		nft_use_dec(&set->use);
+		return;
+	...
+}
+```
+
+Then, when the gc is executed in `nft_set_commit_update`, `nft_trans_gc_catchall_sync` calls `nft_setelem_data_deactivate` to deactivate the objects mapped to the set element. As a result, the `nft_chain` or `nft_object` mapped to the set element can be deactivated twice.
```

Comment:
Can you comment here more about how the reference counter changes of the `nft_chain` object?

Let's consider the `chain_leak`: you use for the `immediate` expr, so the ref. counter becomes 1, then you set up the anonymous set and the lookup expression for the chain (ref count becomes 2), and then you wait for the timeout (ref count becomes 1) and then how the lookup expression is destroyed? Is it during the `del_chain` call?

If yes, then the ref count becomes 0 at this point but the `immediate` expr is still referencing the `nft_chain` object?

Reply:
The set element in the victim set that the lookup expr is binding to, not the lookup expr, is referencing the chain. If the gc is executed while deleting the lookup expr (deleting the lookup expr deactivates the bound set if it is an anonymous set), the set element will be deactivated twice, decreasing the reference count of the victim chain to twice.

I added a description about this to the `exploit.md`.